export_snowflake/__init__.py

```diff
@@ -0,0 +1,175 @@
+
+    Params:
+        config: configuration dictionary
+
+    Returns:
+        tuple of retrieved items: table_cache, file_format_type
+    """
+    db = DbSync(config)
+    # The file format is detected at DbSync init time
+    file_format_type = db.file_format.file_format_type
+    
+    return file_format_type
+
+def direct_transfer_data_from_s3_to_snowflake(config, o, file_format_type):    
+    # retrieve the schema file from s3 bucket
+    # if the table doesn't exist, create the new table based on the schema
+    s3_client = boto3.client('s3')
+
+    bucket = config["bucket"]
+    prefix = config["prefix"]
+    response = s3_client.list_objects_v2(Bucket=bucket, Prefix=prefix)
+
+    schema_file = None
+    for obj in response['Contents']:
+        # for both python and spark parquet_to_csv steps, we always have a schema.json with the schema details
+        if obj['Key'].endswith('.json'):
+            schema_file = obj['Key']
+    
+    if schema_file is None:
+        raise Exception("Schema file not found in the s3 bucket")
+
+    s3_client.download_file(bucket, schema_file, LOCAL_SCHEMA_FILE_PATH)
+
+    f = open(LOCAL_SCHEMA_FILE_PATH)
+    schema = json.load(f)
+
+    # generate the schema object for DbSync
+    stream = config["stream"]
+    o = {
+        "stream": stream,
+        "schema": {"properties": schema['properties']},
+        "key_properties": config["key_columns"]
+    }
+    
+    try:
+        transfer_start_time = time.time()
+
+        # check if the schema(Snowflake table) exists and create it if not
+        db_sync = DbSync(config, o, None, file_format_type)
+        db_sync.create_schema_if_not_exists()
+        
+        # sync_table will check the schema with the table in Snowflake, if there's a miss matching in columns, raise an error
+        db_sync.sync_table()
+        
+        # generate a new stage with stream in Snowflake
+        # the stage will be an external stage that points to the s3
+        # the following merge query will be processed directly against the external stage
+        db_sync.generate_temporary_external_s3_stage(bucket, prefix, config['s3_credentials'])
+        
+        # after creating the external stage, we could load the file directly from the s3 to Snowflake
+        # need to specify the patterns in the s3 bucket to filter out the target csv.gz files
+        db_sync.load_file()
+        
+        transfer_end_time = time.time()
+        stream = config["stream"]
+        LOGGER.info(f"Elapsed time usage for {stream} is {transfer_end_time - transfer_start_time}")
+    except Exception as e:
+        LOGGER.error(f"Error occurred in direct_transfer_data_from_s3_to_snowflake: {e}")
+        raise e
+    finally:
+        # remove the schema file from local
+        os.remove(LOCAL_SCHEMA_FILE_PATH)
+        
+        # Snowflake will only remove the external stage object, the s3 bucket and files will remain
+        db_sync.remove_external_s3_stage()
```

Comment:
Is write privilege enough to create external stage in sf?

Reply:
I tested with the privileges we list in the Snowflake Export doc.
To export to an existing Snowflake table, we used to ask for privileges on 
`USAGE` on `WAREHOUSE, DATABASE, SCHEMA, FILE FORMAT` and `OWNERSHIP` on `TABLE XX`
This will be sufficient for creating an external stage.

It's not specified, but I think the schema privilege will be sufficient for creating the external stages.